tinytuya/core.py

```diff
@@ -333,13 +333,22 @@ def hex2bin(x):
 
 def set_debug(toggle=True, color=True):
     """Enable tinytuya verbose logging"""
+    log._tinytuya_use_color = color
+
+    if color:
+        # \x1b[39;2m
+        fmt = "\x1b[0m\x1b[31;1m%(levelname)s:\x1b[22m%(name)s:\x1b[31;1m%(message)s\x1b[0m"

```

Comment:
I haven't tried, but wonder if we could leverage more colorama here instead to make it more readable. I mean, it's fine.

Reply:
Yeah, I originally wanted to but couldn't decide between `a + b + '%(string)' + c +d` (ugh) or doing `%s%s%s%s%s%s` while trying to get the % formatting to not mangle the embedded %'s (also ugh).  I'll see if I can find a more-readable way that I'm happy with.

Reply:
Ok, I think the choice is between
`fmt = ''.join( Style.RESET_ALL, Fore.RED, Style.BRIGHT, '%(levelname)s:', Style.NORMAL, '%(name)s:', Fore.RED, Style.BRIGHT, '%(message)s', Style.RESET_ALL )`
and
`fmt = Style.RESET_ALL + Fore.RED + Style.BRIGHT + '%(levelname)s:' + Style.NORMAL + '%(name)s:' + Fore.RED + Style.BRIGHT + '%(message)s' + Style.RESET_ALL`

Any preference?

Reply:
I prefer the second one (addition).  Very simple.  Also, I can't get the first one to work in my test.

TypeError: str.join() takes exactly one argument (10 given)

Reply:
Ah, I forgot the parentheses around the list, should have been `fmt = ''.join( (...) )`.  Oh well, we're using the additive one anyway.

While I was at it I rewrote `termcolor()`  as well:
```python
def termcolor(color=True):
    if color is False:
        # Disable Terminal Color Formatting
        bold = subbold = normal = dim = alert = alertdim = cyan = red = yellow = ""
    else:
        # Terminal Color Formatting
        bold     = Style.RESET_ALL + Fore.LIGHTWHITE_EX + Style.BRIGHT
        subbold  = Style.RESET_ALL + Fore.GREEN
        normal   = Style.RESET_ALL
        dim      = Style.RESET_ALL + Fore.LIGHTWHITE_EX + Style.DIM
        alert    = Style.RESET_ALL + Fore.LIGHTRED_EX + Style.BRIGHT
        alertdim = Style.RESET_ALL + Fore.LIGHTRED_EX + Style.DIM
        cyan     = Style.RESET_ALL + Fore.CYAN
        red      = Style.RESET_ALL + Fore.RED
        yellow   = Style.RESET_ALL + Fore.YELLOW
    return bold,subbold,normal,dim,alert,alertdim,cyan,red,yellow
```

Reply:
For join, we would need to make it into an array.  I saw some post that join can be faster than addition, but it wasn't apparent to me and I still think the addition is cleaner:

```python
from colorama import Fore, Style
import time 

print("\nJoin Method: ")
start = time.time()
for x in range(200):
    fmt = ''.join([Style.RESET_ALL, Fore.RED, Style.BRIGHT, '%(levelname)s:', Style.NORMAL, '%(name)s:', Fore.RED, Style.BRIGHT, '%(message)s', Style.RESET_ALL])
end = time.time()
print("The time difference is :", end - start)

print("\nAddition Method: ")
start = time.time()
for x in range(200):
    fmt = (Style.RESET_ALL + Fore.RED + Style.BRIGHT + '%(levelname)s:' + Style.NORMAL + '%(name)s:' + Fore.RED + Style.BRIGHT + '%(message)s' + Style.RESET_ALL)
end = time.time()
print("The time difference is :", end - start)
```

Reply:
Nice! Yes, love the `termcolor()` update!

Reply:
Upping the loop count to 200,000 makes it a bit more obvious, but the difference is so minuscule it's not worth worrying about.  I don't really care either way and currently have it as:
```python
    if color:
        fmt = Style.RESET_ALL + Fore.RED + Style.BRIGHT + '%(levelname)s:' + Style.NORMAL + '%(name)s:' + Fore.RED + Style.BRIGHT + '%(message)s' + Style.RESET_ALL
    else:
        fmt = "%(levelname)s:%(name)s:%(message)s"
```